tools/manifest/sourcefile.py

```diff
@@ -868,6 +868,61 @@ def type(self):
         rv, _ = self.manifest_items()
         return rv
 
+    @property
+    def possible_types(self):
```

Comment:
Hmm, so I don't love duplicating the logic across multiple functions; although I appreciate the assert to check that we're not getting inconsistent output.

Reply:
Do you have any proposal about how to avoid the duplication?

Reply:
Not that's good enough to actually propose. I was thinking you could have something like

```py
def manifest_items(self):
  types, multiple_tests = self.possible_types()
  if len(types) > 1:
      this_type = self.get_definite_type()
  else:
     this_type = types.pop()
  if multiple_tests:
    tests = this.get_all_items()
  else:
    args, kwargs = this.args_by_type(type)
    tests = [this_type(*args, **kwargs)]
  return this_type.item_name, tests
```

Reply:
That still just ends up with duplicate between `possible_types` and `get_definite_type`, no? It also totally fails to deal with the case where the URL needs to differ or variants, etc.